app/services/physical_infra_topology_service.rb

```diff
@@ -78,16 +100,16 @@ def entity_status(entity)
     end
   end
 
-  # Decide whether or not to add an entity to the graph stack
-  def add_to_graph?(entity, links_index, links, parent_id)
-    return true if entity.kind_of?(Tag) || entity.kind_of?(PhysicalSwitch)
-
-    idx = links_index[entity_id(entity)]
-    # If a node has already been processed, change its parent rather than pushing to the stack
-    if idx
-      links[idx][:source] = parent_id
-      return false
+  # We don't allow a node to have more than one parent.
+  # The Physical Infra Topology must be a Tree without double links to the same node.
+  def filter_links(links)
+    has_parent = {}
+    links.reverse.each_with_object([]) do |link, filtered_links|
+      target = link[:target]
+      unless has_parent[target]
```

Comment:
@martinpovolny I couldn't use the `return` inside this block. It breaks.

Reply:
Yes, I fixed a bug in my comment `s/return/next` but never mind.